File: compilers/rjs/transpiler/parser/src/ast/expression.rs
use std::borrow::Cow;
use serde::{Deserialize, Serialize};
use crate::ast::declaration::{ClassBody, FunctionBody};
use crate::span::Span;
use rjs_attribute_marco::js_expr_node;

use super::declaration::{ClassAccessor, ClassConstructor, ClassMethodDefinition};
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct SuperExpression;
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ThisExpression;
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct Identifier<'a> {
    pub name: Cow<'a, str>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct PrivateName<'a> {
    pub name: Cow<'a, str>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct NumberLiteral<'a> {
    pub raw_val: Cow<'a, str>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct StringLiteral<'a> {
    pub raw_val: Cow<'a, str>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct BoolLiteral<'a> {
    pub raw_val: Cow<'a, str>,
    pub value: bool,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct NullLiteral;
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct UndefinedLiteral;
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct RegexLiteral<'a> {
    pub pattern: Cow<'a, str>,
    pub flags: Cow<'a, str>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct TemplateLiteral<'a> {
    pub quasis: Vec<TemplateElement<'a>>,
    pub exprs: Vec<Expression<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct TemplateElement<'a> {
    pub raw_val: Cow<'a, str>,
    pub tail: bool,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ObjectExpression<'a> {
    pub properties: Vec<PropertyDefinition<'a>>,
    pub trailing_comma: bool,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(tag="type")]
pub enum PropertyName<'a> {
    #[serde(rename="Identifier")]
    Ident(Identifier<'a>),
    #[serde(rename="StringLiteral")]
    String(StringLiteral<'a>),
    #[serde(rename="NumberLiteral")]
    Number(NumberLiteral<'a>),
    #[serde(untagged)]
    Expr(Expression<'a>),
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(tag="type")]
pub enum PropertyDefinition<'a> {
    #[serde(rename="ObjectProperty")]
    ObjProp(ObjectProperty<'a>),
    #[serde(rename="ObjectMethodDefinition")]
    ObjMethodDef(ObjectMethodDefinition<'a>),
    #[serde(rename="SpreadElement")]
    Spread(SpreadElement<'a>),
    #[serde(rename="ObjectAccessor")]
    ObjAccessor(ObjectAccessor<'a>)
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ObjectProperty<'a> {
    pub key: PropertyName<'a>,
    pub value: Option<Expression<'a>>,
    pub computed: bool,
    pub shorted: bool,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ObjectMethodDefinition<'a> {
    pub key: PropertyName<'a>,
    pub body: FunctionBody<'a>,
    pub params: Vec<Pattern<'a>>,
    pub computed: bool,
    pub generator: bool,
    pub is_async: bool,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum MethodDefinition<'a> {
    ObjMethod(ObjectMethodDefinition<'a>),
    ClassMethod(ClassMethodDefinition<'a>),
    ObjAccessor(ObjectAccessor<'a>),
    ClassCtor(ClassConstructor<'a>),
    ClassAccessor(ClassAccessor<'a>),
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ObjectAccessor<'a> {
    pub key: PropertyName<'a>,
    pub accessor_type: AccessorType,
    pub body: FunctionBody<'a>,
    pub params: Vec<Pattern<'a>>,
    pub computed: bool,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct SpreadElement<'a> {
    pub argument: Box<Expression<'a>>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum AccessorType {
    Setter,
    Getter
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ClassExpression<'a> {
    pub id: Option<Identifier<'a>>,
    pub super_class: Option<Box<Expression<'a>>>,
    pub body: ClassBody<'a>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ArrayExpression<'a> {
    pub elements: Vec<Option<Expression<'a>>>,
    pub trailing_comma: bool,
}
#[js_expr_node]
pub struct FunctionExpression<'a> {
    pub id: Option<Identifier<'a>>,
    pub params: Vec<Pattern<'a>>,
    pub body: FunctionBody<'a>,
    pub generator: bool,
    pub is_async: bool,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ArrowFunctionExpression<'a> {
    pub is_async: bool,
    pub expression_body: bool,
    pub body: ExpressionOrFunctionBody<'a>,
    pub params: Vec<Pattern<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum ExpressionOrFunctionBody<'a> {
    Expr(Box<Expression<'a>>),
    FuncBody(FunctionBody<'a>)
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct MetaPropery<'a> {
    pub meta: Identifier<'a>,
    pub property: Identifier<'a>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct AwaitExpression<'a> {
    pub argument: Box<Expression<'a>>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct NewExpression<'a> {
    pub callee: Box<Expression<'a>>,
    pub arguments: Vec<Expression<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct MemberExpression<'a> {
    pub object: Box<Expression<'a>>,
    pub property: Box<Expression<'a>>,
    pub computed: bool,
    pub optional: bool
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct CallExpression<'a> {
    pub callee: Box<Expression<'a>>,
    pub arguments: Vec<Expression<'a>>,
    pub optional: bool
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct TaggedTemplateExpression<'a> {
    pub quasi: TemplateLiteral<'a>,
    pub tag: Box<Expression<'a>>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ChainExpression<'a> {
    pub expr: Box<Expression<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct UpdateExpression<'a> {
    pub argument: Box<Expression<'a>>,
    pub prefix: bool,
    pub operator:UpdateOperatorKinds,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum UpdateOperatorKinds {
    IncreOperator,
    DecreOperator
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct UnaryExpression<'a> {
    pub argument: Box<Expression<'a>>,
    pub operator: UnaryOperatorKinds,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum UnaryOperatorKinds {
    LogicalNOTOperator, // !
    BitwiseNOTOperator, // ~
    BitwiseXOROperator, // ^
    PlusOperator,       // +
    MinusOperator,      // -
    DeleteKeyword,      // delete
    VoidKeyword,        // void
    TypeofKeyword       // typeof
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct BinaryExpression<'a> {
    pub left: Box<Expression<'a>>,
    pub right: Box<Expression<'a>>,
    pub operator: BinaryOperatorKinds
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum BinaryOperatorKinds {
    PlusOperator,       // +
    MinusOperator,      // -
    DivideOperator,     // /
    MultiplyOperator,   // *
    ModOperator,    // %
    ExponOperator,  // **
    GtOperator,     // >
    LtOperator,     // <
    EqOperator,     // ==
    NotEqOperator,  // !=
    GeqtOperator,   // >=
    LeqtOperator,   // <=
    StrictEqOperator,       // ===
    StrictNotEqOperator,    // !==
    BitwiseOROperator,      // |
    BitwiseANDOperator,     // &
    BitwiseXOROperator,     // ^
    BitwiseLeftShiftOperator,      // <<
    BitwiseRightShiftOperator,     // >>
    BitwiseRightShiftFillOperator,  // >>>
    LogicalANDOperator, // &&
    LogicalOROperator, // ||
    InKeyword,
    InstanceofKeyword,
    NullishOperator // ??
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ConditionalExpression<'a> {
    pub test: Box<Expression<'a>>,
    pub conseq: Box<Expression<'a>>,
    pub alter: Box<Expression<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct YieldExpression<'a> {
    pub argument: Option<Box<Expression<'a>>>,
    pub deletgate: bool,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct AssignmentExpression<'a> {
    pub left: Box<Pattern<'a>>,
    pub right: Box<Expression<'a>>,
    pub operator: AssignmentOperatorKinds
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub enum AssignmentOperatorKinds {
   AssginOperator,
   PlusAssignOperator,
   MinusAssignOperator,
   ExponAssignOperator,
   DivideAssignOperator,
   MultiplyAssignOperator,
   ModAssignOperator,
   BitwiseORAssginOperator,    // |=
   BitwiseANDAssginOperator,   // &=
   BitwiseNOTAssginOperator,   // ~=
   BitwiseXORAssginOperator,   // ^=
   LogicalORAssignOperator ,   // ||=
   LogicalAndassginOperator,   // &&=
   BitwiseLeftShiftAssginOperator,     // <<=
   BitwiseRightShiftAssginOperator,    // >>=
   BitwiseRightShiftFillAssginOperator // >>>=
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct SequenceExpression<'a> {
    pub exprs: Vec<Expression<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(tag="type")]
pub enum Expression<'a> {
    // idents
    #[serde(rename="SuperExpression")]
    Super(SuperExpression),
    #[serde(rename="ThisExpression")]
    This(ThisExpression),
    #[serde(rename="Identifier")]
    Ident(Identifier<'a>),
    #[serde(rename="PrivateName")]
    Private(PrivateName<'a>),
    // literals
    #[serde(rename="NumberLiteral")]
    Number(NumberLiteral<'a>),
    #[serde(rename="StringLiteral")]
    String(StringLiteral<'a>),
    #[serde(rename="BoolLiteral")]
    Bool(BoolLiteral<'a>),
    #[serde(rename="TemplateLiteral")]
    Template(TemplateLiteral<'a>),
    #[serde(rename="UndefinedLiteral")]
    Undefined(UndefinedLiteral),
    #[serde(rename="NullLiteral")]
    Null(NullLiteral),
    #[serde(rename="ObjectExpression")]
    ObjectExpr(ObjectExpression<'a>), 
    #[serde(rename="ArrayExpression")]
    ArrayExpr(ArrayExpression<'a>),
    #[serde(rename="ArrowFunctionExpression")]
    ArrorFunctionExpr(ArrowFunctionExpression<'a>),
    #[serde(rename="FunctionExpression")]
    FunctionExpr(FunctionExpression<'a>),
    #[serde(rename="ClassExpression")]
    ClassExpr(ClassExpression<'a>),
    // meta
    #[serde(rename="SpreadElement")]
    Spread(SpreadElement<'a>),
    #[serde(rename="MetaProperty")]
    Meta(MetaPropery<'a>),
    // exprs
    #[serde(rename="CallExpression")]
    CallExpr(CallExpression<'a>),
    #[serde(rename="MemberExpression")]
    MemberExpr(MemberExpression<'a>),
    #[serde(rename="TaggedTemplateExpression")]
    TaggedTemplateExpr(TaggedTemplateExpression<'a>),
    #[serde(rename="NewExpression")]
    NewExpr(NewExpression<'a>),
    #[serde(rename="ChainExpression")]
    ChainExpr(ChainExpression<'a>),
    #[serde(rename="UpdateExpression")]
    UpdateExpr(UpdateExpression<'a>),
    #[serde(rename="UnaryExpression")]
    UnaryExpr(UnaryExpression<'a>),
    #[serde(rename="AwaitExpression")]
    AwaitExpr(AwaitExpression<'a>),
    #[serde(rename="BinaryExpression")]
    BinaryExpr(BinaryExpression<'a>),
    #[serde(rename="ConditionalExpression")]
    ConditionalExpr(ConditionalExpression<'a>),
    #[serde(rename="YieldExpression")]
    YieldExpr(YieldExpression<'a>),
    #[serde(rename="AssignmentExpression")]
    AssigmentExpr(AssignmentExpression<'a>),
    #[serde(rename="SequenceExpression")]
    SequenceExpr(SequenceExpression<'a>),
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(tag="type")]
pub enum Pattern<'a> {
    #[serde(rename="Identifier")]
    Ident(Identifier<'a>),
    #[serde(rename="AssignmentPattern")]
    Assgin(AssignmentPattern<'a>),
    #[serde(rename="ObjectPattern")]
    Obj(ObjectPattern<'a>),
    #[serde(rename="ArrayPattern")]
    Array(ArrayPattern<'a>),
    #[serde(rename="RestElement")]
    Rest(RestElement<'a>),
    #[serde(rename="MemberExpression")]
    MemberExpr(MemberExpression<'a>)
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ObjectPattern<'a> {
    pub properties: Vec<ObjectPatternProperty<'a>>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
#[serde(tag="type")]
pub enum  ObjectPatternProperty<'a> {
    #[serde(rename="ObjectProperty")]
    Property {
        key: PropertyName<'a>,
        value: Option<Box<Pattern<'a>>>,
        computed: bool,
        shorted: bool,
    },
    #[serde(rename="AssignmentPattern")]
    Assign(AssignmentPattern<'a>),
    #[serde(rename="RestElement")]
    Rest(RestElement<'a>),
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct AssignmentPattern<'a> {
    pub left: Box<Pattern<'a>>,
    pub right: Expression<'a>,
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct ArrayPattern<'a> {
    pub elements: Vec<Option<Pattern<'a>>>
}
#[derive(Debug, Clone, PartialEq, Deserialize, Serialize)]
pub struct RestElement<'a> {
    pub argument: Box<Pattern<'a>>,
}

pub trait ExprNodeParan {
    fn set_paran(&mut self, is_paran: bool);
}

File: web-infras/common/src/ast/expression.ts
import {
  AssigmentOperatorKinds,
  BinaryOperatorKinds,
  SyntaxKinds,
  UnaryOperatorKinds,
  UpdateOperatorKinds,
} from "@/src/kind";
import { ExpressionModuleItem, ModuleItem } from "@/src/ast/base";
import { Pattern } from "@/src/ast/pattern";
import { PropertyName, FunctionBody, Function, Class } from "@/src/ast/declaration";
import {
  TSTypeAnnotation,
  TSTypeNode,
  TSTypeParameterDeclaration,
  TSTypeParameterInstantiation,
} from "@/src/ast/typescript";

export interface Super extends ExpressionModuleItem {
  kind: SyntaxKinds.Super;
  name: "super";
}
export interface Import extends ExpressionModuleItem {
  kind: SyntaxKinds.Import;
  name: "import";
}
export interface ThisExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.ThisExpression;
  name: "this";
}
export interface Identifier extends ExpressionModuleItem {
  kind: SyntaxKinds.Identifier;
  name: string;
  typeAnnotation: TSTypeAnnotation | undefined;
  optional: boolean | undefined;
}
export interface PrivateName extends ExpressionModuleItem {
  kind: SyntaxKinds.PrivateName;
  name: string;
}
export interface DecimalLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.DecimalLiteral;
  rawValue: string;
}
export interface DecimalBigIntegerLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.DecimalBigIntegerLiteral;
  rawValue: string;
}
export interface NonOctalDecimalLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.NonOctalDecimalLiteral;
  rawValue: string;
}
export interface BinaryIntegerLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.BinaryIntegerLiteral;
  rawValue: string;
}
export interface OctalIntegerLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.OctalIntegerLiteral;
  rawValue: string;
}
export interface HexIntegerLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.HexIntegerLiteral;
  rawValue: string;
}
export interface BinaryBigIntegerLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.BinaryBigIntegerLiteral;
  rawValue: string;
}
export interface OctalBigIntegerLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.OctalBigIntegerLiteral;
  rawValue: string;
}
export interface HexBigIntegerLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.HexBigIntegerLiteral;
  rawValue: string;
}
export interface LegacyOctalIntegerLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.LegacyOctalIntegerLiteral;
  rawValue: string;
}
export type NumberLiteral =
  | DecimalLiteral
  | NonOctalDecimalLiteral
  | BinaryIntegerLiteral
  | OctalIntegerLiteral
  | HexIntegerLiteral
  | LegacyOctalIntegerLiteral
  | BinaryBigIntegerLiteral
  | HexBigIntegerLiteral
  | OctalBigIntegerLiteral
  | DecimalBigIntegerLiteral;
export interface StringLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.StringLiteral;
  value: string;
}
export interface BoolLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.BooleanLiteral;
  value: boolean;
}
export interface NullLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.NullLiteral;
}
export interface UndefinbedLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.UndefinedLiteral;
}
export interface RegexLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.RegexLiteral;
  pattern: string;
  flag: string;
}
export interface TemplateLiteral extends ExpressionModuleItem {
  kind: SyntaxKinds.TemplateLiteral;
  quasis: Array<TemplateElement>;
  expressions: Array<Expression>;
}
export interface TemplateElement extends ExpressionModuleItem {
  kind: SyntaxKinds.TemplateElement;
  value: string;
  tail: boolean;
}
export interface ObjectExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.ObjectExpression;
  properties: Array<PropertyDefinition>;
  trailingComma: boolean;
}
export type PropertyDefinition = ObjectProperty | ObjectMethodDefinition | SpreadElement | ObjectAccessor;
export interface ObjectProperty extends ModuleItem {
  kind: SyntaxKinds.ObjectProperty;
  key: PropertyName;
  value: Expression | undefined; // actually is assignment expression,
  computed: boolean;
  shorted: boolean;
}
export interface ObjectMethodDefinition extends ModuleItem {
  kind: SyntaxKinds.ObjectMethodDefintion;
  key: PropertyName;
  body: FunctionBody;
  params: Array<Pattern>;
  typeParameters: TSTypeParameterDeclaration | undefined;
  returnType: TSTypeAnnotation | undefined;
  computed: boolean;
  generator: boolean;
  async: boolean;
}
export interface ObjectAccessor extends ModuleItem {
  kind: SyntaxKinds.ObjectAccessor;
  type: "get" | "set";
  key: PropertyName | PrivateName;
  body: FunctionBody;
  params: Array<Pattern>;
  typeParameters: TSTypeParameterDeclaration | undefined;
  returnType: TSTypeAnnotation | undefined;
  computed: boolean;
}
export interface SpreadElement extends ExpressionModuleItem {
  kind: SyntaxKinds.SpreadElement;
  argument: Expression;
}
export interface ClassExpression extends Class, ExpressionModuleItem {
  kind: SyntaxKinds.ClassExpression;
}
export interface ArrayExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.ArrayExpression;
  elements: Array<Expression | null>; // actually need to be assigment expression;
  trailingComma: boolean;
}
export interface FunctionExpression extends ExpressionModuleItem, Function {
  kind: SyntaxKinds.FunctionExpression;
}
// TODO: make arrowfunctionExpression extends form function
export interface ArrorFunctionExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.ArrowFunctionExpression;
  expressionBody: boolean;
  async: boolean;
  arguments: Array<Pattern>;
  typeParameters: TSTypeParameterDeclaration | undefined;
  returnType: TSTypeAnnotation | undefined;
  body: Expression | FunctionBody;
}
export interface MetaProperty extends ExpressionModuleItem {
  kind: SyntaxKinds.MetaProperty;
  meta: Identifier | Import;
  property: Identifier;
}
export interface AwaitExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.AwaitExpression;
  argument: Expression; // actually is unary expression
}
export interface NewExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.NewExpression;
  callee: Expression;
  arguments: Array<Expression>;
  typeArguments: TSTypeParameterInstantiation | undefined;
}
export interface MemberExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.MemberExpression;
  object: Expression;
  property: Expression | PrivateName;
  computed: boolean;
  optional: boolean;
}
export interface CallExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.CallExpression;
  callee: Expression;
  arguments: Array<Expression>;
  typeArguments: TSTypeParameterInstantiation | undefined;
  optional: boolean;
}
export interface TaggedTemplateExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.TaggedTemplateExpression;
  quasi: TemplateLiteral;
  tag: Expression;
}
export interface ChainExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.ChainExpression;
  expression: Expression;
}
export interface UpdateExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.UpdateExpression;
  argument: Expression;
  prefix: boolean;
  operator: UpdateOperatorKinds;
}
export interface UnaryExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.UnaryExpression;
  argument: Expression;
  operator: UnaryOperatorKinds;
}
export interface BinaryExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.BinaryExpression;
  left: Expression;
  right: Expression;
  operator: BinaryOperatorKinds;
}
export interface ConditionalExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.ConditionalExpression;
  test: Expression;
  consequnce: Expression;
  alter: Expression;
}
export interface YieldExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.YieldExpression;
  argument: Expression | null;
  delegate: boolean;
}
export interface AssigmentExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.AssigmentExpression;
  left: Pattern;
  right: Expression;
  operator: AssigmentOperatorKinds;
}
export interface SequenceExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.SequenceExpression;
  exprs: Array<Expression>;
}

export interface TSInstantiationExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.TSInstantiationExpression;
  typeArguments: TSTypeParameterInstantiation;
  expression: Expression;
}

export interface TSTypeAssertionExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.TSTypeAssertionExpression;
  expression: Expression;
  typeAnnotation: TSTypeNode;
}

export interface TSAsExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.TSAsExpression;
  expression: Expression;
  typeAnnotation: TSTypeNode;
}

export interface TSSatisfiesExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.TSSatisfiesExpression;
  expression: Expression;
  typeAnnotation: TSTypeNode;
}

export interface TSNonNullExpression extends ExpressionModuleItem {
  kind: SyntaxKinds.TSNonNullExpression;
  expression: Expression;
}

export type Expression =
  // jsx element
  | JSXElement
  | JSXFragment
  // identifer and super and ThisExpression
  | Identifier
  | PrivateName
  | Super
  | ThisExpression
  | Import
  // literals
  | NumberLiteral
  | StringLiteral
  | BoolLiteral
  | TemplateLiteral
  | UndefinbedLiteral
  | NullLiteral
  | RegexLiteral
  // structal literal
  | ObjectExpression
  | ArrayExpression
  | ArrorFunctionExpression
  | FunctionExpression
  | ClassExpression
  // meta property and spread element
  | SpreadElement
  | MetaProperty
  // other expression
  | CallExpression
  | MemberExpression
  | TaggedTemplateExpression
  | NewExpression
  | ChainExpression
  | UpdateExpression
  | UnaryExpression
  | AwaitExpression
  | BinaryExpression
  | ConditionalExpression
  | YieldExpression
  | AssigmentExpression
  | SequenceExpression
  // TS expression
  | TSInstantiationExpression
  | TSTypeAssertionExpression
  | TSAsExpression
  | TSSatisfiesExpression
  | TSNonNullExpression;

/** ====================================
 *  JSX Element or JSX fragment
 * =====================================
 */
export interface JSXElement extends ExpressionModuleItem {
  kind: SyntaxKinds.JSXElement;
  openingElement: JSXOpeningElement;
  children: Array<JSXText | JSXExpressionContainer | JSXSpreadChild | JSXElement | JSXFragment>;
  closingElement: JSXClosingElement | null;
}
export interface JSXIdentifier extends ModuleItem {
  kind: SyntaxKinds.JSXIdentifier;
  name: string;
}
export interface JSXMemberExpression extends ModuleItem {
  kind: SyntaxKinds.JSXMemberExpression;
  object: JSXMemberExpression | JSXIdentifier;
  property: JSXIdentifier;
}
export interface JSXNamespacedName extends ModuleItem {
  kind: SyntaxKinds.JSXNamespaceName;
  namespace: JSXIdentifier;
  name: JSXIdentifier;
}
export interface JSXAttribute extends ModuleItem {
  kind: SyntaxKinds.JSXAttribute;
  name: JSXIdentifier | JSXNamespacedName;
  value:
    | NumberLiteral
    | StringLiteral
    | BoolLiteral
    | TemplateLiteral
    | UndefinbedLiteral
    | NullLiteral
    | RegexLiteral
    | JSXExpressionContainer
    | JSXElement
    | JSXFragment
    | null;
}
export interface JSXSpreadAttribute extends ModuleItem {
  kind: SyntaxKinds.JSXSpreadAttribute;
  argument: Expression;
}
export interface JSXOpeningElement extends ModuleItem {
  kind: SyntaxKinds.JSXOpeningElement;
  name: JSXIdentifier | JSXMemberExpression | JSXNamespacedName;
  attributes: Array<JSXAttribute | JSXSpreadAttribute>;
  selfClosing: boolean;
}
export interface JSXClosingElement extends ModuleItem {
  kind: SyntaxKinds.JSXClosingElement;
  name: JSXIdentifier | JSXMemberExpression | JSXNamespacedName;
}
export interface JSXSpreadChild extends ModuleItem {
  kind: SyntaxKinds.JSXSpreadChild;
  argument: Expression;
}
export interface JSXText extends ModuleItem {
  kind: SyntaxKinds.JSXText;
  value: string;
}
export interface JSXExpressionContainer extends ModuleItem {
  kind: SyntaxKinds.JSXExpressionContainer;
  expression: Expression | null;
}
export interface JSXFragment extends ExpressionModuleItem {
  kind: SyntaxKinds.JSXFragment;
  openingFragment: JSXOpeningFragment;
  children: Array<JSXText | JSXExpressionContainer | JSXSpreadChild | JSXElement | JSXFragment>;
  closingFragment: JSXClosingFragment;
}
export interface JSXOpeningFragment extends ModuleItem {
  kind: SyntaxKinds.JSXOpeningFragment;
}
export interface JSXClosingFragment extends ModuleItem {
  kind: SyntaxKinds.JSXClosingFragment;
}
